flytekit/extras/persistence/s3_awscli.py

```diff
@@ -0,0 +1,162 @@
+            retry += 1
+            if retry > _aws_config.RETRIES.get():
+                raise
+            secs = _aws_config.BACKOFF_SECONDS.get()
+            logging.info(f"Sleeping before retrying again, after {secs} seconds")
+            time.sleep(secs)
+            logging.info("Retrying again")
+
+
+def _extra_args(extra_args: Dict[str, str]) -> List[str]:
+    cmd = []
+    if "ContentType" in extra_args:
+        cmd += ["--content-type", extra_args["ContentType"]]
+    if "ContentEncoding" in extra_args:
+        cmd += ["--content-encoding", extra_args["ContentEncoding"]]
+    if "ACL" in extra_args:
+        cmd += ["--acl", extra_args["ACL"]]
+    return cmd
+
+
+class S3Persistence(DataPersistence):
+    """
+    DataPersistence plugin for AWS S3 (and Minio). Use aws cli to manage the transfer. The binary needs to be installed
+    separately
+
+    .. prompt::
+
+       pip install awscli
+
+    """
+
+    PROTOCOL = "s3://"
+    _AWS_CLI = "aws"
+    _SHARD_CHARACTERS = [str(x) for x in range(10)] + list(_string.ascii_lowercase)
+
+    def __init__(self, default_prefix: Optional[str] = None):
+        super().__init__(name="awscli-s3", default_prefix=default_prefix)
+
+    @staticmethod
+    def _check_binary():
+        """
+        Make sure that the AWS cli is present
+        """
+        if not shell_which(S3Persistence._AWS_CLI):
+            raise _FlyteUserException("AWS CLI not found! Please install it with `pip install awscli`.")
+
+    @staticmethod
+    def _split_s3_path_to_bucket_and_key(path: str) -> (str, str):
+        """
+        splits a valid s3 uri into bucket and key
+        """
+        path = path[len("s3://") :]
+        first_slash = path.index("/")
+        return path[:first_slash], path[first_slash + 1 :]
+
+    def exists(self, remote_path):
+        """
+        Given a remote path of the format s3://, checks if the remote file exists
+        """
+        S3Persistence._check_binary()
+
+        if not remote_path.startswith("s3://"):
+            raise ValueError("Not an S3 ARN. Please use FQN (S3 ARN) of the format s3://...")
+
+        bucket, file_path = self._split_s3_path_to_bucket_and_key(remote_path)
+        cmd = [
+            S3Persistence._AWS_CLI,
+            "s3api",
+            "head-object",
+            "--bucket",
+            bucket,
+            "--key",
+            file_path,
+        ]
+        try:
+            _update_cmd_config_and_execute(cmd)
+            return True
+        except Exception as ex:
+            # The s3api command returns an error if the object does not exist. The error message contains
+            # the http status code: "An error occurred (404) when calling the HeadObject operation: Not Found"
+            #  This is a best effort for returning if the object does not exist by searching
+            # for existence of (404) in the error message. This should not be needed when we get off the cli and use lib
```

Comment:
is the plan to replace the cli with fsspec using s3fs? or a boto implementation?

Reply:
we cannot replace, as there are existing users of this. but users can optionally install `pip install flytekitplugins-data-fsspec` or other such solutions in the future